h/presenters/annotation_json.py

```diff
@@ -9,68 +9,59 @@
 
 
 class AnnotationJSONPresenter(AnnotationBasePresenter):
-
     """Present an annotation in the JSON format returned by API requests."""
 
     def __init__(self, annotation_context, formatters=None):
         super().__init__(annotation_context)
 
-        self._formatters = []
-
-        if formatters is not None:
-            self._formatters.extend(formatters)
+        self._formatters = tuple(formatters or [])
 
     def asdict(self):
-        docpresenter = DocumentJSONPresenter(self.annotation.document)
-
-        base = {
-            "id": self.annotation.id,
-            "created": self.created,
-            "updated": self.updated,
-            "user": self.annotation.userid,
-            "uri": self.annotation.target_uri,
-            "text": self.text,
-            "tags": self.tags,
-            "group": self.annotation.groupid,
-            "permissions": self.permissions,
-            "target": self.target,
-            "document": docpresenter.asdict(),
-            "links": self.links,
-        }
+        annotation = deepcopy(self.annotation.extra) or {}
+
+        annotation.update(
+            {
+                "id": self.annotation.id,
+                "created": self.created,
+                "updated": self.updated,
+                "user": self.annotation.userid,
+                "uri": self.annotation.target_uri,
+                "text": self.text,
+                "tags": self.tags,
+                "group": self.annotation.groupid,
+                #  Convert our simple internal annotation storage format into the
+                #  legacy complex permissions dict format that is still used in
+                #  some places.
+                "permissions": {
+                    "read": [self._get_read_permission()],
+                    "admin": [self.annotation.userid],
+                    "update": [self.annotation.userid],
+                    "delete": [self.annotation.userid],
+                },
+                "target": self.target,
+                "document": DocumentJSONPresenter(self.annotation.document).asdict(),
+                "links": self.links,
+            }
+        )
```

Comment:
This is the same stuff except we start from the cloned extras and update it instead of creating a separate dict and doing that after. This made more sense to me as it means the code is laid out in the order the data is applied.

The permissions are also set here, because only one of them changes, and as far as I can tell this property is only ever used locally.

Reply:
Yeah it seems easier to read here. Removes a local variable that was just making things harder to follow